src/window.rs

```diff
@@ -60,46 +59,45 @@ impl Window {
         };
 
         if self.frame_count == 0 {
+            // NOTE(mbernat): It's possible to avoid initial mode setting
+            // since we're keeping the previous mode: we could just call page_flip directly.
             self.drm_display.set_mode_with_framebuffer(Some(fb));
         } else {
             self.drm_display.page_flip(fb);
         }
+
+        // NOTE(mbernat): This buffer object returns to the surface's queue upon destruction.
+        // If we were to release it here, it would be again available from
+        // `Surface::lock_front_buffer()` next and the app would be effectively single-buffered.
+        // So, we keep it around for one frame, which is fine for double buffering.
+        self.previous_bo.replace(buffer_object);
+
         self.frame_count += 1;
     }
 
     pub fn restore_original_display(&self) {
-        self.drm_display.set_mode_with_framebuffer(self.drm_display.crtc.framebuffer());
+        let handle = self.drm_display.crtc.framebuffer().unwrap();
+        if self.drm_display.gbm_device.get_framebuffer(handle).is_ok() {
+            self.drm_display.set_mode_with_framebuffer(self.drm_display.crtc.framebuffer());
+        }
     }
 
     pub fn draw(&mut self, mut drawer: impl FnMut()) {
+        drawer();
+        // SAFETY: eglSwapBuffers is called by `frame.finish()` in drawer()
+        unsafe { self.present() };
+
         // The first page flip is scheduled after frame #1 (which is the second frame)
         // Yes, this is very stupid, just testing if it works
 
         if self.frame_count > 1 {
-            let mut events =
+            let _events =
                 self.drm_display.gbm_device.receive_events().expect("Could not receive events");
```

Comment:
Is `receive_events()` a blocking call? I'm not sure what the consequences of calling it are if we're not past frame 1.

Reply:
Yes, it's blocking. If you haven't scheduled page flips or vblanks previously you get an infinite loop here, which is arguably terrible. And using frame numbers for this logic is really opaque, we should track this in some state enum, I'll make the change here.

Also, the underlying implementation is just a select on the DRM file descriptor and a decent implementation would select from multiple sources (e.g. keyboard), like the C impls do. I'm not sure the DRM crate gives us this option but it will be necessary if we want to support an event loop later.

Reply:
Ohhh I see now, it's not doing much more than just reading from the file descriptor. That's good in a way, we could potentially add our own implementation which allows for timeouts or other behavior.